core/src/server/rpc_module.rs

```diff
@@ -877,63 +966,64 @@ impl SubscriptionSink {
 		.map_err(Into::into)
 	}
 
+	fn build_error_message<T: Serialize>(&self, error: &T) -> Result<String, Error> {
+		serde_json::to_string(&SubscriptionError::new(
+			self.method.into(),
+			SubscriptionPayloadError { subscription: self.uniq_sub.sub_id.clone(), error },
+		))
+		.map_err(Into::into)
+	}
+
 	fn inner_send(&mut self, msg: String) -> Result<(), Error> {
 		let res = match self.is_connected.as_ref() {
 			Some(conn) if !conn.is_canceled() => {
 				// unbounded send only fails if the receiver has been dropped.
-				self.inner.send_raw(msg).map_err(|_| Some(SubscriptionClosedReason::ConnectionReset))
+				self.inner.send_raw(msg).map_err(|_| Some(SubscriptionClosed::ConnectionReset))
 			}
-			Some(_) => Err(Some(SubscriptionClosedReason::Unsubscribed)),
+			Some(_) => Err(Some(SubscriptionClosed::Unsubscribed)),
 			// NOTE(niklasad1): this should be unreachable, after the first error is detected the subscription is closed.
 			None => Err(None),
 		};
 
-		// The subscription was already closed by the client
-		// Close down the subscription but don't send a message to the client.
-		if res.is_err() {
-			self.inner_close(None);
-		}
-
 		res.map_err(|e| {
-			let err = e.unwrap_or_else(|| SubscriptionClosedReason::Server("Close reason unknown".to_string()));
+			let err = e.unwrap_or_else(|| CloseReason::Unknown.into());
 			Error::SubscriptionClosed(err.into())
 		})
 	}
 
-	/// Close the subscription sink with a customized error message.
-	pub fn close_with_custom_message(&mut self, msg: &str) {
-		let close_reason = SubscriptionClosedReason::Server(msg.to_string()).into();
-		self.inner_close(Some(&close_reason));
-	}
-
-	/// Close the subscription sink with the provided [`SubscriptionClosed`].
-	pub fn close(&mut self, close_reason: &SubscriptionClosed) {
-		self.inner_close(Some(close_reason));
-	}
-
-	fn inner_close(&mut self, maybe_close: Option<&SubscriptionClosed>) {
+	/// Send an error notification on the subscription with a special field `error`.
+	///
+	/// This may be used for indicating to connected clients that a given subscription has been canceled
+	/// as well so not necessarily an error but this is used for compat reason with existing JavaScript libraries.
+	///
+	/// The APIs [`SubscriptionSink::pipe_from_stream`] and [`SubscriptionSink::pipe_from_stream`] uses this
+	/// when the stream is terminated.
+	///
+	/// ```json
+	/// {
+	///  "jsonrpc": "2.0",
+	///  "method": "<method>",
+	///  "params": {
+	///    "subscription": "<subscriptionID>",
+	///    "error": { "code": <your code>, "message": <your message>, "data": <your data but might be omitted> }
+	///    }
+	///  }
+	/// }
+	/// ```
+	///
+	pub fn close(&mut self, err: Error) -> bool {
```

Comment:
Out of interest, is there a reason that this needs to take `&mut self` and not `mut self` (so that once it's been closed, people can't try using the sink again)?

You've added code so that an `Error` can be converted into an `ErrorObject`, but I don't think one can convert from an `ErrorObject` to an `Error` as easily. When somebody has an error, do we want people to create a Jsonrpsee `Error` (eg a `CallError`) normally, or would we prefer them to reach for `ErrorObject`?

If we want them to use the jsonrpsee `Error` type (eg `CallError`) then maybe we shouldn't really expose functions that take `ErrorObject`, and just let them construct `Error`s instead, which we internally convert (if people want full control, they can use a `CallError::Custom` by the looks of it)?

Reply:
> Out of interest, is there a reason that this needs to take &mut self and not mut self (so that once it's been closed, people can't try using the sink again)?

good catch, it makes sense to take ownership there.

> You've added code so that an Error can be converted into an ErrorObject, but I don't think one can convert from an ErrorObject to an Error as easily. When somebody has an error, do we want people to create a Jsonrpsee Error (eg a CallError) normally, or would we prefer them to reach for ErrorObject?

ErrorObject(borrowed) and CallError::Custom(owned) is the same thing, but `CallError::Custom` take ownership of `data.
We could could extract a take `ErrorObjectOwned` and add it in `CallError` maybe this more understandable?

`CallError` is used by the servers and `ErrorObject` by clients.

I guess if we do want I said :point_up: then it should make sense right?

Reply:
https://github.com/paritytech/jsonrpsee/pull/728/commits/17ef15b8296f789c6dd7cba6e5f3a9afaf0ff2b2